### DISCUSSION TITLE
How to add a global output filter on text nodes?

### DISCUSSION THREAD

**Discussion by bplace** [thephpleague/commonmark#641] [Q&A] [answered]
## Question

Hi, I've been struggling to properly implement a simple filter on all text nodes. 
I want to replace the space before all colons (":") characters by a non-breaking space ("&nbsp;"). 

Quite straightforward with a regular expression, but I can't modify the string content with a Listener. I've been messing around inline or block renderers, but it looks like I would have to make a custom renderer for every block.

Is there a way to achieve this easily?

**Comment by colinodell** [comment]
Maybe you could try adding an [event listener](https://commonmark.thephpleague.com/1.5/customization/event-dispatcher/#example) that responds to the `DocumentParsedEvent` event, looks for the `Text` nodes, and then does that regex replacement on them.  This would ensure the filtering is done after everything is parsed and you wouldn't need to override any core components.

Here's some quick and dirty code I wrote without my IDE that _might_ have bugs, but should get you close:

```php
<?php
class AddNonBreakingSpacesBeforeColonsProcessor
{
    public function onDocumentParsed(DocumentParsedEvent $event)
    {
        $document = $event->getDocument();
        $walker = $document->walker();
        while ($event = $walker->next()) {
            $node = $event->getNode();

            if (!($node instanceof Text) || !$event->isEntering()) {
                continue;
            }

            $content = $node->getContent();
            // TODO: Perform your regex find-and-replace here on $content
            $node->setContent($content);
        }
    }
}
```

Then add it to your `Environment`:

```php
$environment->addEventListener(DocumentParsedEvent::class, [new AddNonBreakingSpacesBeforeColonsProcessor(), 'onDocumentParsed']);
```

I hope that helps!

**Comment by bplace** [comment]
Hi Colin, thank you for answering. I missed this Element class, which is quite obvious.

I'm not sure I can use it though, as the text node content get escaped in the renderer, so I would need to split the text node in two, with an additional Inline node between. Looks difficult to do.

I will try to add a pre renderer on the whole markdown input, or have a custom Document renderer. Will post my findings here.

**Comment by bplace** [comment, marked as answer]
Here's what I did:

**In the Extension:**

```
$environment->addBlockRenderer(Document::class, new TypographyRenderer(), 10);
```

**TypographyRenderer**


```
    public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, bool $inTightList = false)
    {
        if ( ! ($block instanceof Document)) {
            throw new \InvalidArgumentException('Incompatible block type: ' . \get_class($block));
        }

        $wholeDoc = $htmlRenderer->renderBlocks($block->children());

        $wholeDoc = preg_replace("#(\w) :#", "$1<span class=\"u-fine\">&nbsp;</span>:", $wholeDoc);

        return $wholeDoc === '' ? '' : $wholeDoc . "\n";
    }
```

Any other little tweak I'll need will go there, gotta be careful on the regular expression though!
